test/variant_sender_test.cpp

```diff
@@ -0,0 +1,204 @@
+
+  EXPECT_TRUE(!!result);
+  EXPECT_EQ(*result, 5);
+
+  try {
+      auto just_error_variant_sender = func(false);
+      EXPECT_FALSE(just_error_variant_sender.sends_done);
+      sync_wait(just_error_variant_sender);
+      EXPECT_FALSE(true);
+  } catch (int& v) {
+      EXPECT_EQ(v, 10);
+  }
+
+  std::cout << "variant_sender done " << *result << "\n";
+}
+
+TEST(Variant, CombineJustAndDone) {
+  auto func = [](bool v) -> variant_sender<decltype(just(5)), decltype(just_done())> {
+      if (v) {
+          return just(5);
+      } else {
+          return just_done();
+      }
+  };
+
+  auto just_variant_sender = func(true);
+  EXPECT_TRUE(just_variant_sender.sends_done);
+  auto result = sync_wait(just_variant_sender);
+
+  EXPECT_TRUE(!!result);
+  EXPECT_EQ(*result, 5);
+
+  auto just_done_variant_sender = func(false);
+  EXPECT_TRUE(just_done_variant_sender.sends_done);
+  auto result2 = sync_wait(just_done_variant_sender);
+  EXPECT_FALSE(!!result2);
+
+  std::cout << "variant_sender done " << *result << "\n";
+}
+
+TEST(Variant, CombineJustAndJust) {
+  auto func = [&](bool v) -> variant_sender<decltype(just(5)), decltype(dematerialize(materialize(just(42))))> {
+      if (v) {
+          return just(5);
+      } else {
+          return dematerialize(materialize(just(42)));
+      }
+  };
+
+  auto just_variant_sender = func(true);
+  EXPECT_FALSE(just_variant_sender.sends_done);
+  auto result = sync_wait(just_variant_sender);
+
+  EXPECT_TRUE(!!result);
+  EXPECT_EQ(*result, 5);
+
+  auto materialized_variant_sender = func(false);
+  auto result2 = sync_wait(materialized_variant_sender);
+  EXPECT_TRUE(!!result2);
+  EXPECT_EQ(*result2, 42);
+
+  std::cout << "variant_sender done " << *result << "\n";
+}
+
+TEST(Variant, CombineJustAndJust_Invalid) {
+  auto func = [](bool v) -> variant_sender<decltype(just(5)), decltype(just(std::declval<std::string>()))> {
+      if (v) {
+          return just(5);
+      } else {
+          return just(std::string("Hello World"));
+      }
+  };
+
+  IntAndStringReceiver rec;
+
+  auto just_variant_sender = func(true);
+  EXPECT_FALSE(just_variant_sender.sends_done);
+  auto op = unifex::connect(just_variant_sender, rec);
+  unifex::start(op);
+
+  auto just_string_sender = func(false);
+  EXPECT_FALSE(just_variant_sender.sends_done);
+  auto op2 = unifex::connect(just_string_sender, rec);
+  unifex::start(op2);
+}
+
+template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept>
+using test_sender_t = variant_sender<TestSender<lvalueConnectNoexcept, rvalueConnectNoexcept>>;
+
+template<typename T, bool lvalue>
+using conditionally_lvalue_t = std::conditional_t<lvalue, std::add_lvalue_reference_t<T>, T>;
+
+template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept, bool isLvalueReference = true>
+using is_noexcept = unifex::is_nothrow_connectable<conditionally_lvalue_t<test_sender_t<lvalueConnectNoexcept, rvalueConnectNoexcept>, isLvalueReference>, IntAndStringReceiver>;
```

Comment:
These should be in an anonymous namespace, I think.
```suggestion
namespace {

template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept>
using test_sender_t = variant_sender<TestSender<lvalueConnectNoexcept, rvalueConnectNoexcept>>;

template<typename T, bool lvalue>
using conditionally_lvalue_t = std::conditional_t<lvalue, std::add_lvalue_reference_t<T>, T>;

template<bool lvalueConnectNoexcept, bool rvalueConnectNoexcept, bool isLvalueReference = true>
using is_noexcept = unifex::is_nothrow_connectable<conditionally_lvalue_t<test_sender_t<lvalueConnectNoexcept, rvalueConnectNoexcept>, isLvalueReference>, IntAndStringReceiver>;

} // namespace
```